+++
weight = 0
hero = "midjourney4.png"
pubmed = ""
doi = "10.1101/2022.04.23.489288 "
date = 2022-04-24T13:36:35-07:00
title = "Growth phase estimation for abundant bacterial populations sampled longitudinally from human stool metagenomes"
+++

Longitudinal sampling of the stool has yielded important insights into the ecological dynamics of the human gut microbiome. However, due to practical limitations, the most densely sampled time series from the human gut are collected at a frequency of about once per day, while the population doubling times for gut commensals are on the order of minutes-to-hours. Despite this, much of the prior work on human gut microbiome time series modeling has, implicitly or explicitly, assumed that day-to-day fluctuations in taxon abundances are related to population growth or death rates, which is likely not the case. Here, we propose an alternative model of the human gut as a continuous flow ecosystem at a dynamical steady state, where population dynamics occur internally and the bacterial population sizes measured in stool represent an endpoint of these internal dynamics. We formalize this idea as stochastic logistic growth of a population held at a constant dilution rate. We show how this model provides a path toward estimating the growth phases of gut bacterial populations in situ. We assess our model predictions against densely-sampled human stool metagenomic time series data. Consistent with our model, donors with slower defecation rates tended to harbor a larger proportion of taxa in later growth phases, while faster defecation rates were associated with more taxa in earlier growth phases. We discuss how these growth phase estimates may be used to better inform metabolic modeling in flow-through ecosystems, like animal guts or industrial bioreactors.

{{< tweet user="gibbological" id="1518628359630061568" >}}
